### ISSUE TITLE
Checkstyle xpath suppressions fails when the suppressions file is empty

### ISSUE THREAD

**Issue by Fernal73** [checkstyle/checkstyle#12987]
> 
> This sounds like a new issue. I don't know for sure, but I can only assume we expected the user to look at the generated result before trying to use it so it displays nothing if it found nothing and a single suppression if it does and not append to the existing suppression file.

I'd expect checkstyle to check if the suppressions file is empty or not before attempting to parse the document. If the suppressions weren't being generated to the console, I'd expect checkstyle to generate the suppressions document including the xml header and empty suppressions list when no suppressions are found to be generated. 
In my usecase, it makes sense to delete the existing suppressions file before generating fresh ones since I'd expect to retain the existing ones plus the new ones if found. In either case, an empty suppressions file will always crash checkstyle as it currently is.

_Originally posted by @Fernal73 in https://github.com/checkstyle/checkstyle/issues/12969#issuecomment-1506064048_
            
Kindly run ant main and ant main2 on the repo https://github.com/Fernal73/CheckStyleBug to reproduce the error in two different processes.

**Comment by Fernal73** [comment]
```bash
/var/tmp $ javac ./src/com/github/sevntu/checkstyle/Sample.java

/var/tmp $ cat checkstyle.xml
<?xml version="1.0"?>
<!DOCTYPE
module PUBLIC "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
"https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="SuppressionXpathFilter">
      <property name="optional" value="true" />
      <property name="file" value="suppressions-xpath.xml" />
    </module>
  </module>
</module>

/var/tmp $ cat ./src/com/github/sevntu/checkstyle/Sample.java
package com.github.sevntu.checkstyle;

public class Sample {
  public Boolean shouldBeWarning() {
    return Boolean.TRUE;
  }
}

Scenario I:
/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.9.3-all.jar -c checkstyle.xml -g -o suppressions-xpath.xml  ./src/com/github/sevntu/checkstyle/Sample.java
com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module TreeWalker - cannot initialize module SuppressionXpathFilter - Unable to parse suppressions-xpath.xml - Premature end of file.
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:480)
        at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:207)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:411)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:338)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:195)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:130)
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module SuppressionXpathFilter - Unable to parse suppressions-xpath.xml - Premature end of file.
        at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:128)
        at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:207)
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:475)
        ... 5 more
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Unable to parse suppressions-xpath.xml - Premature end of file.
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getSuppressionLoader(SuppressionsLoader.java:292)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadXpathSuppressions(SuppressionsLoader.java:267)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadXpathSuppressions(SuppressionsLoader.java:253)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionXpathFilter.finishLocalSetup(SuppressionXpathFilter.java:583)
        at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:203)
        at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:123)
        ... 7 more
Caused by: org.xml.sax.SAXParseException; systemId: file:/root/CheckStyleBug/suppressions-xpath.xml; lineNumber: 1; columnNumber: 1; Premature end of file.
        at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:204)
        at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.fatalError(ErrorHandlerWrapper.java:178)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:400)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLScanner.reportFatalError(XMLScanner.java:1465)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:1013)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)
        at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)
        at com.puppycrawl.tools.checkstyle.XmlLoader.parseInputSource(XmlLoader.java:81)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getSuppressionLoader(SuppressionsLoader.java:283)
        ... 12 more
Checkstyle ends with 1 errors.

Scenario II:
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.9.3-all.jar -c checktyle.xml -g ./src/com/github/sevntu/checkstyle/Sample.java 2> gen.err 1>suppressions-xpath.xml

/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.9.3-all.jar -c checkstyle.xml ./src/com/github/sevntu/checkstyle/Sample.java
 
com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module TreeWalker - cannot initialize module SuppressionXpathFilter - Unable to parse suppressions-xpath.xml - Premature end of file.
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:480)
        at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:207)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:411)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:338)
        at com.puppycrawl.tools.checkstyle.Main.execute(Main.java:195)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:130)
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module SuppressionXpathFilter - Unable to parse suppressions-xpath.xml - Premature end of file.
        at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:128)
        at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:207)
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:475)
        ... 5 more
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Unable to parse suppressions-xpath.xml - Premature end of file.
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getSuppressionLoader(SuppressionsLoader.java:292)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadXpathSuppressions(SuppressionsLoader.java:267)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.loadXpathSuppressions(SuppressionsLoader.java:253)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionXpathFilter.finishLocalSetup(SuppressionXpathFilter.java:583)
        at com.puppycrawl.tools.checkstyle.AbstractAutomaticBean.configure(AbstractAutomaticBean.java:203)
        at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:123)
        ... 7 more
Caused by: org.xml.sax.SAXParseException; systemId: file:/root/CheckStyleBug/suppressions-xpath.xml; lineNumber: 1; columnNumber: 1; Premature end of file.
        at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.createSAXParseException(ErrorHandlerWrapper.java:204)
        at java.xml/com.sun.org.apache.xerces.internal.util.ErrorHandlerWrapper.fatalError(ErrorHandlerWrapper.java:178)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:400)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLErrorReporter.reportError(XMLErrorReporter.java:327)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLScanner.reportFatalError(XMLScanner.java:1465)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl$PrologDriver.next(XMLDocumentScannerImpl.java:1013)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentScannerImpl.next(XMLDocumentScannerImpl.java:605)
        at java.xml/com.sun.org.apache.xerces.internal.impl.XMLDocumentFragmentScannerImpl.scanDocument(XMLDocumentFragmentScannerImpl.java:542)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:889)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XML11Configuration.parse(XML11Configuration.java:825)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.XMLParser.parse(XMLParser.java:141)
        at java.xml/com.sun.org.apache.xerces.internal.parsers.AbstractSAXParser.parse(AbstractSAXParser.java:1224)
        at java.xml/com.sun.org.apache.xerces.internal.jaxp.SAXParserImpl$JAXPSAXParser.parse(SAXParserImpl.java:637)
        at com.puppycrawl.tools.checkstyle.XmlLoader.parseInputSource(XmlLoader.java:81)
        at com.puppycrawl.tools.checkstyle.filters.SuppressionsLoader.getSuppressionLoader(SuppressionsLoader.java:283)
        ... 12 more
Checkstyle ends with 1 errors.
```


---

**Describe what you expect in detail.**
No RuntimeException should be thrown.

**Comment by strkkk** [comment]
@Fernal73 what behavior do you expect in this case? It seems like exception is the only good thing to do. Silent ignore of bad suppressions file does not seem like a good alternative to me.

**Comment by romani** [comment]
I think that we have here not clear expectations.
Checkstyle is a library, not a final solution.
Through a time we got CLI implementation that was intended for testing and some demo usage.
We have report that is limited to report violations on files. So we do not know how to report this problem, as this is configuration problem.

We need to review our code to see if we already can report such problems in supported formats: plain, xml, ... .

If not if this format can handle general problems, we should stay with raw exception

**Comment by Fernal73** [comment]
I think 

> @Fernal73 what behavior do you expect in this case? It seems like exception is the only good thing to do. Silent ignore of bad suppressions file does not seem like a good alternative to me.

I think we can follow the convention of the checkstyle xml report and generate an empty list of suppressions when there are no suppressions found. The problem is that the empty file generated by no suppressions discoovered is not acceptable to checkstyle.

**Comment by Fernal73** [comment]
> > @Fernal73 what behavior do you expect in this case? It seems like exception is the only good thing to do. Silent ignore of bad suppressions file does not seem like a good alternative to me.
> 
> I think we can follow the convention of the checkstyle xml report and generate an empty list of suppressions when there are no suppressions found. The problem is that the empty file generated by no suppressions discovered is not acceptable to checkstyle.

Additionally, in Scenario I, checkstyle throws an exception even if a non-empty suppressions file exists and the same suppressions file is designated as the output file using the -o cli parameter. Assuming that this can be fixed, what should be the behaviour in this case? Should only incremental suppressions be generated or should the output include the existing suppressions? A compromise could be that, in this scenario, the merged suppressions should be output. In all other cases, only the incremental ones discovered. This would have to be documented as such for the cli.

**Comment by Fernal73** [comment]
> > > @Fernal73 what behavior do you expect in this case? It seems like exception is the only good thing to do. Silent ignore of bad suppressions file does not seem like a good alternative to me.
> > 
> > 
> > I think we can follow the convention of the checkstyle xml report and generate an empty list of suppressions when there are no suppressions found. The problem is that the empty file generated by no suppressions discovered is not acceptable to checkstyle.
> 
> Additionally, in Scenario I, checkstyle throws an exception even if a non-empty suppressions file exists and the same suppressions file is designated as the output file using the -o cli parameter. Assuming that this can be fixed, what should be the behaviour in this case? Should only incremental suppressions be generated or should the output include the existing suppressions? A compromise could be that, in this scenario, the merged suppressions should be output. In all other cases, only the incremental ones discovered. This would have to be documented as such for the cli.

Another way of doing this is for Checkstyle to ignore existing suppressions file while generating suppressions and generate a  fresh and complete list of suppressions. This is cleaner, IMO.

**Comment by romani** [comment]
> The problem is that the empty file generated by no suppressions discovered is not acceptable to checkstyle.

This is valid defect on our side.

**Comment by Fernal73** [comment]
> > The problem is that the empty file generated by no suppressions discovered is not acceptable to checkstyle.
> 
> This is valid defect on our side.
> 
> we will not put in output existing suppressions, as there numerous places where they exeists and they might be not easy to access. CLI is not final solution it just a tool to do some work. Human with brain still required to operate it.

I am not suggesting any more command line options because I don't wish to clutter the cli.

**Comment by Fernal73** [comment]
> > The problem is that the empty file generated by no suppressions discovered is not acceptable to checkstyle.
> 
> This is valid defect on our side.
> 
> we will not put in output existing suppressions, as there numerous places where they exeists and they might be not easy to access. CLI is not final solution it just a tool to do some work. Human with brain still required to operate it.

What do you mean by this?

**Comment by Fernal73** [comment]
> > > The problem is that the empty file generated by no suppressions discovered is not acceptable to checkstyle.
> > 
> > 
> > This is valid defect on our side.
> > IMO, it is not a valid output for -o parameter. For stdout, yes.
> > we will not put in output existing suppressions, as there numerous places where they exeists and they might be not easy to access. CLI is not final solution it just a tool to do some work. Human with brain still required to operate it.
> 
> I meant only the xpath suppressions. No other.

Checkstyle can't and shouldn't provide an output that's an invalid input to itself.

**Comment by Fernal73** [comment]
Also discussed [here...](https://github.com/checkstyle/checkstyle/issues/8418#issuecomment-660155538)
